util/conditions/experimental/summary.go

```diff
@@ -0,0 +1,129 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package experimental
+
+import (
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+// SummaryOption is some configuration that modifies options for a summary request.
+type SummaryOption interface {
+	// ApplyToSummary applies this configuration to the given summary options.
+	ApplyToSummary(*SummaryOptions)
+}
+
+// SummaryOptions allows to set options for the summary operation.
+type SummaryOptions struct {
+	mergeStrategy                  MergeStrategy
+	conditionTypes                 []string
+	negativePolarityConditionTypes []string
+	stepCounter                    bool
+}
+
+// ApplyOptions applies the given list options on these options,
+// and then returns itself (for convenient chaining).
+func (o *SummaryOptions) ApplyOptions(opts []SummaryOption) *SummaryOptions {
+	for _, opt := range opts {
+		opt.ApplyToSummary(o)
+	}
+	return o
+}
+
+// NewSummaryCondition creates a new condition by summarizing a set of conditions from an object.
+// If any of the condition in scope does not exist in the source object, missing conditions are considered Unknown, reason NotYetReported.
+//
+// By default, the Aggregate condition has the same type of the source condition, but this can be changed by using
+// the OverrideType option.
+//
+// Additionally, it is possible to inject custom merge strategies using the WithMergeStrategy option or
+// to add a step counter to the generated message by using the WithStepCounter option.
+func NewSummaryCondition(obj runtime.Object, conditionType string, opts ...SummaryOption) (*metav1.Condition, error) {
```

Comment:
```suggestion
func NewSummaryCondition(obj runtime.Object, targetConditionTYpe string, opts ...SummaryOption) (*metav1.Condition, error) {
```
maybe

Maybe it would be good if we go over all utils and call all obj & conditionTypes one of:
* sourceObj / sourceObjs
* targetObj / targetObjs
* sourceConditionType
* targetConditionType

But only in the cases where the parameters are always either source or target. I know there are cases where the passed in conditionTypes can be both source and target if there is no override

Reply:
Fixed